### PR TITLE
Admin store create

### LINKED ISSUE #9: admin/store/create GET POST handlers
The store/create GET and POST handlers need to be created for the admin user.

A skeleton for the GET handler is available [here](https://github.com/blueimpact/kucipong/blob/master/src/Kucipong/Handler/Admin.hs#L64).  It should be relatively simple.  We may not really want to to do anything else with it until we figure out how we are doing the templating (#7).

The POST handler needs to do the following things:
- authenticate the admin user
  - you can look at how the GET handler is doing authentication
- receive the email address for the store user
  - parse this from one of the query params
- create a new `CompanyEmail` row in the database
  - add a new class method to `MonadKucipongDb`
- send the new store user a registration email
  - This is similar to the one we send the admin after adding a new admin.
- redirect the admin to a success (or error) page
  - The `login` handler for admins is doing a redirect, so you can look at how they are doing it.

@arowM Feel free to work on either the GET or POST handler.

### PR DESCRIPTION
**Pull request by cdepillabout**
This PR implements the admin store-create POST handler.

This PR also updates to lts-7 (ghc-8.0.1) in order to use the new http-api-data package for easily parsing http form data.

@arowM Please review.  Especially the `Kucipong.Db.Models.EntityDefs` module.  Let me know what you think.

Closes #9.  Closes #15.

### CODE REVIEW

File: src/Kucipong/Handler/Admin.hs [hunk 1]
```diff
@@ -4,90 +4,119 @@ module Kucipong.Handler.Admin where
...
+storeCreatePost = do
+    (AdminSession email) <- getAdminEmail
+    (AdminStoreCreateForm storeEmailParam) <- getReqParam
+    (Entity storeEmailKey storeEmail) <- dbCreateStoreEmail storeEmailParam
+    (Entity storeLoginTokenKey storeLoginToken) <-
+        dbCreateStoreMagicLoginToken storeEmailKey
+    sendStoreLoginEmail
+        (storeEmailEmail storeEmail)
+        (storeLoginTokenLoginToken storeLoginToken)
+    -- TODO: Where to redirect this?
```
Comment (arowM) [inline src/Kucipong/Handler/Admin.hs]: Redirect to the store create page, please.
I'll implement to show user as a pop up message on the page that the user has successfully created the store user, later.

File: src/Kucipong/Spock/ReqParam.hs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+getReqParam
+    :: forall ctx m a
+     . (FromForm a, MonadIO m)
+    => ActionCtxT ctx m a
+getReqParam = getReqParamErr handleFormDecodeError
+  where
+    handleFormDecodeError :: Text -> ActionCtxT ctx m a
+    handleFormDecodeError err = do
+        -- TODO: How should we handle this error?
```
Comment (arowM) [inline src/Kucipong/Spock/ReqParam.hs:27]: Please make it possible to render any page template we want with errors as template parameters.
I'll implement later to show the errors as pop up message on the page.

**Review by arowM — CHANGES_REQUESTED** [review]
It's good generally.
But I've not still clearly realized why we use two tables (`Store` and `StoreEmail`) as store and not for admin user.

The followings are my comments for the questions in the code.

**Comment by cdepillabout** [comment]
@arowM Thanks for the review.

> I've not still clearly realized why we use two tables (`Store` and `StoreEmail`) as store and not for admin user.

When creating the `Admin` user, all the fields are known (both the email and the name of the admin user).  So there is no reason to split up the `Admin` table into two different tables.

But when creating the Store user, only the email is known.  So we split up the information about a store into two different tables, `StoreEmail` and `Store`.  Store users that have been created in the database, but not actually logged in only exist in the `StoreEmail`.

As soon as the store user actually actually logs in and setups their store, then they also exist in the `Store` table.  Some of the fields in the `Store` table are required (like the name of the store, the category, etc).  But some fields are not required (like an image of the store).

This is similar to how we actually think about stores in real life.  It makes sense to only register emails for stores (but no other information) before the `Store` user actually logs in and sets up their store.  However, after the `Store` user logs in, it doesn't really make sense to have a `Store` with a name that can be `null`.  

My goal was to model the store tables similar to our real-life expectations about them.

However, it's entirely possible that this makes the `Store` table hard to actually use.  If we find out this is the case, then we should change them to your original design.

File: src/Kucipong/Handler/Admin.hs (hunk 1 above)
Comment (cdepillabout) [reply src/Kucipong/Handler/Admin.hs]: Okay, I'll do this (and the next comment) next week :+1:

**Comment by arowM** [comment]
@cdepillabout Thanks for detail explanation.  It makes sense.

File: src/Kucipong/Handler/Admin.hs (hunk 1 above)
Comment (cdepillabout) [reply src/Kucipong/Handler/Admin.hs]: I fixed this in commit 3b49e2e.

File: src/Kucipong/Spock/ReqParam.hs (hunk 2 above)
Comment (cdepillabout) [reply src/Kucipong/Spock/ReqParam.hs:27]: I will make a separate issue about this.

**Comment by cdepillabout** [comment]
I will go ahead and merge this in.

One comment is not yet resolved, but I will create a separate issue about it.

**Pull request merged**